### PR TITLE
Use libsodium based allocator

### LINKED ISSUE #3576: CURVE plaintext should be placed in secure memory
For additional security, unencrypted plaintext should be placed in secure memory when using CURVE security.

Points to consider:
- Is this effective for the message payload? Can we ensure this is for all copies that are under control of libzmq, and enable the user to ensure this as well?
- We should care to reduce the number of secure memory allocations to a minimum since this is a very costly operation.

### PR DESCRIPTION
**Pull request by sigiesec**
Fixes #3576

### CODE REVIEW

**Comment by bluca** [comment]
Nice stuff! Unfortunately there's a couple of build failures, any idea?

```
In file included from ../tests/test_security_curve.cpp:52:0:
../tests/../src/curve_client_tools.hpp:44:2: error: #error "CURVE library not built properly"
 #error "CURVE library not built properly"
```

**Comment by sigiesec** [comment]
Mh, I was struggling to fix the Windows builds but probably broke the other
builds by that.

Will give that another look tomorrow.

Luca Boccassi <notifications@github.com> schrieb am So., 14. Juli 2019,
19:00:

**Comment by sigiesec** [comment]
Strange, I cannot reproduce this problem. This only seems to happen with the builds using a very old gcc...

**Comment by bluca** [comment]
Clang seems to have the same problem: https://travis-ci.org/zeromq/libzmq/jobs/558584750
Include order mixup?

**Comment by sigiesec** [comment]
I think it's fine now :)

File: tests/test_security_curve.cpp [hunk 1]
```diff
@@ -27,6 +27,21 @@
...
+#include "platform.hpp"
+#else
+#include "../src/platform.hpp"
+#endif
+
+#ifndef ZMQ_USE_TWEETNACL
+#define ZMQ_USE_TWEETNACL
+#endif
+#ifdef ZMQ_USE_LIBSODIUM
+#undef ZMQ_USE_LIBSODIUM
```
Comment (bluca) [inline tests/test_security_curve.cpp]: If I follow correctly, this means the unit tests are effectively not using the new memory allocators, right?

**Comment by bluca** [comment]
Nice!

File: tests/test_security_curve.cpp (hunk 1 above)
Comment (sigiesec) [reply tests/test_security_curve.cpp]: Well they are on the side of the tested libzmq functions. Just not for the mock curve client messages

Comment (bluca) [reply tests/test_security_curve.cpp]: Ah yes of course - thanks!

File: src/secure_allocator.hpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        typedef secure_allocator_t<U> other;
+    };
+
+    void construct (pointer p, const_reference val)
+    {
+        new ((void *) p) value_type (val);
+    }
+    void destroy (pointer p) { p->~value_type (); }
+    size_type max_size () const { return SIZE_MAX; }
```
Comment (bluca) [inline src/secure_allocator.hpp]: @sigiesec SIZE_MAX is not available on some old libcs (centos 6, ubuntu 12):

https://build.opensuse.org/package/show/network:messaging:zeromq:git-stable/libzmq

```
[  194s] tests/../src/secure_allocator.hpp: In member function 'zmq::secure_allocator_t<T>::size_type zmq::secure_allocator_t<T>::max_size() const':
[  194s] tests/../src/secure_allocator.hpp:82:42: error: 'SIZE_MAX' was not declared in this scope
```

Could you please add a compat ifnedf in src/stdint?

Comment (bluca) [reply src/secure_allocator.hpp]: Fixed by https://github.com/zeromq/libzmq/pull/3584

**Pull request merged**